### PR TITLE
Add eigenstrain and body force terms to interaction integrals

### PR DESCRIPTION
**Pull request by bwspenc**
ref #18662 

Also addresses a couple of issues with the body force kernels

### CODE REVIEW

File: modules/contact/test/tests/bouncing-block-contact/grid-sequencing/tests [hunk 1]
```diff
@@ -23,5 +23,6 @@
     partial = True
     mesh_mode = replicated # Fix when mortar segment mesh is distributed
     rel_err = 2e-5
+    allow_warnings = True # This is a difficult problem that can result in poor mortar projections that emit warnings
```
Comment (dschwen) [inline modules/contact/test/tests/bouncing-block-contact/grid-sequencing/tests:26]: This seems unrelated, why did you encounter this issue in this context?

File: modules/tensor_mechanics/doc/content/modules/tensor_mechanics/FractureIntegrals.md [hunk 2]
```diff
@@ -70,9 +70,9 @@ To use the $J$-integral capability in MOOSE, a user specifies a set of nodes alo
...
+where $\sigma$ is the stress, $u$ is the displacement, and $q$ is identical to the $q$-functions used for $J$-integrals. The last two terms in this equation account for the effects of the gradient of the eigenstrain, $\epsilon^*$, and for the effects of the body force, $b$. If the eigenstrain is a thermal strain, its gradient is computed using the gradient of the temperature field and the temperature derivative of the eigenstrain. In that case, the user supplies the temperature and eigenstrain name through the `temperature` and `eigenstrain_name` parameters, respectively. If the eigenstrain is from another source, its gradient must be computed by the eigenstrain material model in the form of a RankThreeTensor object, whose name is specified for the fracture integrals using the `eigenstrain_gradient` parameter.
```
Comment (dschwen) [inline modules/tensor_mechanics/doc/content/modules/tensor_mechanics/FractureIntegrals.md:75]:
```suggestion
where $\sigma$ is the stress, $u$ is the displacement, and $q$ is identical to the $q$-functions used for $J$-integrals. The last two terms in this equation account for the effects of the gradient of the eigenstrain, $\epsilon^*$, and for the effects of the body force, $b$. If the eigenstrain is a thermal strain, its gradient is computed using the gradient of the temperature field and the temperature derivative of the eigenstrain. In that case, the user supplies the temperature and eigenstrain name through the [!param](/DomainIntegral/temperature) and [!param](/DomainIntegral/eigenstrain_name) parameters, respectively. If the eigenstrain is from another source, its gradient must be computed by the eigenstrain material model in the form of a RankThreeTensor object, whose name is specified for the fracture integrals using the [!param](/DomainIntegral/eigenstrain_gradient) parameter.
```

We should consistently use `[!param](syntax)`. In the contect of this md it was not immediately clear what object/action these parameters refer to.

As a side note `/DomainIntegral` should really be `/Modules/TensorMechanics/DomainIntegral`!

File: modules/tensor_mechanics/include/kernels/MaterialVectorBodyForce.h [hunk 3]
```diff
@@ -29,8 +29,8 @@ class MaterialVectorBodyForce : public Kernel
   /// coordinate axis this Kernel acts on
   unsigned int _component;
 
-  /// coupled force density vector property
-  const MaterialProperty<RealVectorValue> & _force_density;
+  /// coupled body force vector property
```
Comment (dschwen) [inline modules/tensor_mechanics/include/kernels/MaterialVectorBodyForce.h:32]: I'm not a fan of this change, because as far as I can see the units here are a _force density_ and not a _force_, or am I wrong?

File: modules/tensor_mechanics/src/actions/DomainIntegralAction.C [hunk 4]
```diff
@@ -263,6 +266,24 @@ DomainIntegralAction::DomainIntegralAction(const InputParameters & params)
     if (!poissons_ratio_set)
       _poissons_ratio = getParam<Real>("poissons_ratio");
   }
+
+  if (_integrals.count(J_INTEGRAL) != 0 || _integrals.count(C_INTEGRAL) != 0 ||
```
Comment (dschwen) [inline modules/tensor_mechanics/src/actions/DomainIntegralAction.C:270]: looks like enum are being used instead of enum classes

File: modules/tensor_mechanics/src/vectorpostprocessors/InteractionIntegral.C [hunk 5]
```diff
@@ -125,6 +130,17 @@ InteractionIntegral::InteractionIntegral(const InputParameters & parameters)
   // set unused dimensions to zero
   for (std::size_t i = _ndisp; i < 3; ++i)
     _grad_disp[i] = &_grad_zero;
+
+  if (getParam<MaterialPropertyName>("eigenstrain_gradient") != "")
```
Comment (dschwen) [inline modules/tensor_mechanics/src/vectorpostprocessors/InteractionIntegral.C:134]: are you using this comparison instead of `isParamValid` due to the way the action sets this parameter? Wouldn't this break the standalone use of this class? Or are you relying on the (not guaranteed) behavior that an invalid string parameter defaults to "" automatically (if that's even true).

File: modules/tensor_mechanics/test/include/materials/FunctionIsotropicEigenstrain.h [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "ComputeEigenstrainBase.h"
+
+class FunctionIsotropicEigenstrain : public ComputeEigenstrainBase
```
Comment (dschwen) [inline modules/tensor_mechanics/test/include/materials/FunctionIsotropicEigenstrain.h:14]: Slowly subverting the `Compute...` prefix tradition.

File: modules/tensor_mechanics/test/tests/domain_integral_thermal/interaction_integral_2d.i [hunk 7]
```diff
@@ -69,71 +69,70 @@
 []
 
 [Modules/TensorMechanics/Master]
-  [./master]
+  [all]
     strain = FINITE
     add_variables = true
     incremental = true
```
Comment (dschwen) [inline modules/tensor_mechanics/test/tests/domain_integral_thermal/interaction_integral_2d.i:75]: incremental = true is the default for finite

File: modules/tensor_mechanics/test/tests/domain_integral_thermal/interaction_integral_2d_bf.i [hunk 8]
```diff
@@ -0,0 +1,126 @@
+#This is a regression test that exercises the option to include
+#the effects of the body force in the interaction integral. This
+#uses the same basic model as the other cases in this directory.
+#This is a placeholder until a suitable problem with an analytical
+#solution can be identified.
+
+[GlobalParams]
+  order = FIRST
+  family = LAGRANGE
```
Comment (dschwen) [inline modules/tensor_mechanics/test/tests/domain_integral_thermal/interaction_integral_2d_bf.i:9]: this + `add_variables = true` is maybe a bit of a dicey mix, since the TMA determines the appropriate variable order automatically. FIRST/LAGRANGE is the default for NL and Aux variables anyways, so what block is this applying to, the `DomainIntegralAction` (which is using `std::string` as parameter types for those 👎 )?
As a maintenance task we might want to fix that type of stuff sometime.

**Review by dschwen — COMMENTED** [review]
See above

File: modules/contact/test/tests/bouncing-block-contact/grid-sequencing/tests (hunk 1 above)
Comment (bwspenc) [reply modules/contact/test/tests/bouncing-block-contact/grid-sequencing/tests:26]: Hmm, it looks like I picked up a stray commit somehow. I'll fix that.

File: modules/tensor_mechanics/doc/content/modules/tensor_mechanics/FractureIntegrals.md (hunk 2 above)
Comment (bwspenc) [reply modules/tensor_mechanics/doc/content/modules/tensor_mechanics/FractureIntegrals.md:75]: Thanks, I didn't realize we had a standardized way of doing that -- I'll make that change. I agree about moving this under Modules/TensorMechanics. This was around before that was a thing. Let's save that for later, though!

File: modules/tensor_mechanics/test/include/materials/FunctionIsotropicEigenstrain.h (hunk 6 above)
Comment (bwspenc) [reply modules/tensor_mechanics/test/include/materials/FunctionIsotropicEigenstrain.h:14]: Yeah, I guess so. I've been working on subverting that so long that I didn't even realize I was doing it.

File: modules/tensor_mechanics/include/kernels/MaterialVectorBodyForce.h (hunk 3 above)
Comment (bwspenc) [reply modules/tensor_mechanics/include/kernels/MaterialVectorBodyForce.h:32]: This was mostly just to be consistent with the property name.

Comment (bwspenc) [reply modules/tensor_mechanics/include/kernels/MaterialVectorBodyForce.h:32]: To clarify a bit: The current code in MaterialVectorBodyForce.C has a parameter called `body_force`. I was trying to make everything consistent with that. It is commonly referred to as a body force, even though the understanding is that it is a density. My opinion is that we should stick with that convention. If we want a different name, we should change the name of the parameter for the fracture integral code in this PR as well.

File: modules/tensor_mechanics/src/actions/DomainIntegralAction.C (hunk 4 above)
Comment (bwspenc) [reply modules/tensor_mechanics/src/actions/DomainIntegralAction.C:270]: Yes, that is the case. We should clean that up, but I'd say that should be saved for later.

File: modules/tensor_mechanics/test/tests/domain_integral_thermal/interaction_integral_2d.i (hunk 7 above)
Comment (bwspenc) [reply modules/tensor_mechanics/test/tests/domain_integral_thermal/interaction_integral_2d.i:75]: I'll remove this

File: modules/tensor_mechanics/test/tests/domain_integral_thermal/interaction_integral_2d_bf.i (hunk 8 above)
Comment (bwspenc) [reply modules/tensor_mechanics/test/tests/domain_integral_thermal/interaction_integral_2d_bf.i:9]: I'll just remove order/family from all of the cases affected by this PR.

File: modules/tensor_mechanics/src/vectorpostprocessors/InteractionIntegral.C [hunk 9]
```diff
@@ -125,6 +130,20 @@ InteractionIntegral::InteractionIntegral(const InputParameters & parameters)
   // set unused dimensions to zero
   for (std::size_t i = _ndisp; i < 3; ++i)
     _grad_disp[i] = &_grad_zero;
+
+  if ((isParamValid("eigenstrain_gradient") &&
+       getParam<MaterialPropertyName>("eigenstrain_gradient") !=
+           "")) // This extra check is needed to work with the action
```
Comment (dschwen) [inline modules/tensor_mechanics/src/vectorpostprocessors/InteractionIntegral.C:136]: I'm still seeing this

**Review by dschwen — APPROVED** [review]
Looks good to me now!

File: modules/tensor_mechanics/doc/content/modules/tensor_mechanics/FractureIntegrals.md [hunk 10]
```diff
@@ -70,9 +70,9 @@ To use the $J$-integral capability in MOOSE, a user specifies a set of nodes alo
...
+where $\sigma$ is the stress, $u$ is the displacement, and $q$ is identical to the $q$-functions used for $J$-integrals. The last two terms in this equation account for the effects of the gradient of the eigenstrain, $\epsilon^*$, and for the effects of the body force, $b$. If the eigenstrain is a thermal strain, its gradient is computed using the gradient of the temperature field and the temperature derivative of the eigenstrain. In that case, the user supplies the temperature and eigenstrain name through the [!param](DomainIntegral/temperature) and [!param](DomainIntegral/eigenstrain_name) parameters, respectively. If the eigenstrain is from another source, its gradient must be computed by the eigenstrain material model in the form of a RankThreeTensor object, whose name is specified for the fracture integrals using the [!param](DomainIntegral/eigenstrain_gradient) parameter.
```
Comment (dschwen) [inline modules/tensor_mechanics/doc/content/modules/tensor_mechanics/FractureIntegrals.md:75]:
```suggestion
where $\sigma$ is the stress, $u$ is the displacement, and $q$ is identical to the $q$-functions used for $J$-integrals. The last two terms in this equation account for the effects of the gradient of the eigenstrain, $\epsilon^*$, and for the effects of the body force, $b$. If the eigenstrain is a thermal strain, its gradient is computed using the gradient of the temperature field and the temperature derivative of the eigenstrain. In that case, the user supplies the temperature and eigenstrain name through the [!param](/DomainIntegral/temperature) and [!param]([!param](/DomainIntegral/temperature)/eigenstrain_name) parameters, respectively. If the eigenstrain is from another source, its gradient must be computed by the eigenstrain material model in the form of a RankThreeTensor object, whose name is specified for the fracture integrals using the [!param](/DomainIntegral/eigenstrain_gradient) parameter.
```

**Review by dschwen — CHANGES_REQUESTED** [review]
Missed the leading slashes

File: modules/tensor_mechanics/doc/content/modules/tensor_mechanics/FractureIntegrals.md (hunk 10 above)
Comment (bwspenc) [reply modules/tensor_mechanics/doc/content/modules/tensor_mechanics/FractureIntegrals.md]: Uggh -- fixed!

**Review by dschwen — DISMISSED** [review]
Ok, maybe now :o)

**Comment by bwspenc** [comment]
Cool, that stuff to include the parameters breaks the Mastodon tests:
```
TOKENIZE ERROR: name 'obj_syntax' is not defined
/opt/civet/build_1/moose/modules/tensor_mechanics/doc/content/modules/tensor_mechanics/FractureIntegrals.md:73
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
73│[!param](/DomainIntegral/eigenstrain_name)                                                      │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
```
I wonder if it's missing a path in its documentation configuration or something.

**Comment by dschwen** [comment]
Might as well ping @cbolisetti since this is probably a Mastodon issue.

**Review by dschwen — APPROVED** [review]
Third time's a charm
